src/shape.cpp

```diff
@@ -532,19 +568,23 @@ shape::dynamic_dimension& shape::dynamic_dimension::operator-=(const std::size_t
     assert(this->max >= x);
     this->min -= x;
     this->max -= x;
-    if(this->opt != 0)
-    {
-        assert(this->opt >= x);
-        this->opt -= x;
-    }
+    std::set<std::size_t> new_opts;
+    std::transform(this->opts.begin(),
+                   this->opts.end(),
+                   std::inserter(new_opts, new_opts.begin()),
+                   [&x](const auto& opt) {
+                       assert(opt >= x);
```

Comment:
if `opt < x` then, perhaps return `0` means no optimal value?

Reply:
Opt should be within the range of [min, max]. So the previous assert on `min` would fail first probably.